subsys/llext/llext.c

```diff
@@ -324,3 +324,189 @@ void llext_bootstrap(struct llext *ext, llext_entry_fn_t entry_fn, void *user_da
 		return;
 	}
 }
+
+/*
+ * Prepare a set of extension copies for future restoring. The user has copied
+ * multiple extensions and their dependencies into a flat array. We have to
+ * relink dependency pointers to copies in this array for later restoring. Note,
+ * that all dependencies have to be in this array, if any are missing, restoring
+ * will fail, so we return an error in such cases.
+ */
+int llext_relink_dependency(struct llext *ext, unsigned int n_ext)
+{
+	unsigned int i, j, k;
+
+	for (i = 0; i < n_ext; i++) {
+		for (j = 0; ext[i].dependency[j] && j < ARRAY_SIZE(ext[i].dependency); j++) {
+			for (k = 0; k < n_ext; k++) {
+				if (!strncmp(ext[k].name, ext[i].dependency[j]->name,
+				    sizeof(ext[k].name))) {
+					ext[i].dependency[j] = ext + k;
```

Comment:
Right so rather than directly having a pointer here, can we not say that dependency[j] = k and ditch the pointers altogether?

Reply:
@teburd extensions are usually created one by one. IIUC many use-cases will only have one extension, SOF typically uses several of them for a moderately complex stream, sometimes 2, sometimes 5, and we never destroy them. And there can be multiple streams in the system. So if you want to have them in an array, you'd have to have something like a pool of them, reallocating when it becomes too small... Not sure this would improve anything.